### PR TITLE
compiler: Improve data streaming backend

### PR DESCRIPTION
**Pull request by FabioLuporini**
Yet another bunch of misc compiler tweaks and generalisations, mostly to better support lazy data streaming in PRO

### CODE REVIEW

File: devito/ir/stree/algorithms.py [hunk 1]
```diff
@@ -126,9 +126,8 @@ def stree_build(clusters, profiler=None, **kwargs):
             if i.is_Halo:
                 found = i
             elif i.is_Sync:
-                if profiler._verbosity > 0 or not i.is_async:
-                    attach_section(i)
-                    section = None
+                attach_section(i)
```
Comment (FabioLuporini) [inline devito/ir/stree/algorithms.py:129]: @georgebisbas this is something I had promised in a previous PR

File: devito/arch/archinfo.py [hunk 2]
```diff
@@ -19,6 +19,8 @@
            'get_cuda_path', 'get_hip_path', 'check_cuda_runtime', 'get_m1_llvm_path',
            'Platform', 'Cpu64', 'Intel64', 'IntelSkylake', 'Amd', 'Arm', 'Power',
            'Device', 'NvidiaDevice', 'AmdDevice', 'IntelDevice',
+           # Generic
+           'CPU64',
```
Comment (georgebisbas) [inline devito/arch/archinfo.py:23]: SO both CPU and Cpu ?

File: devito/operator/operator.py [hunk 3]
```diff
@@ -939,21 +955,23 @@ def _emit_apply_profiling(self, args):
             if metrics:
                 perf("Global performance: [%s]" % ', '.join(metrics))
 
+            # Same as above, but excluding the setup phase, e.g. the CPU-GPU
```
Comment (georgebisbas) [inline devito/operator/operator.py:958]: So the data transfer time from device to host is not included ?
is it reported somewhere else?

File: devito/operator/operator.py [hunk 4]
```diff
@@ -975,21 +993,12 @@ def lower_perfentry(v):
...
-
-            itershapes = [",".join(str(i) for i in its) for its in v.itershapes]
-            if len(itershapes) > 1:
-                itershapes = ",".join("<%s>" % i for i in itershapes)
-            elif len(itershapes) == 1:
-                itershapes = itershapes[0]
-            else:
-                itershapes = ""
-            name = "%s%s<%s>" % (k.name, rank, itershapes)
-
```
Comment (georgebisbas) [inline devito/operator/operator.py:992]: As discussed ok

File: devito/tools/abc.py [hunk 5]
```diff
@@ -151,6 +151,22 @@ def __init__(self, a, b, c=4):
 
         return cls(*args, **kwargs)
 
+    def _rreplace(self, subs):
+        """
+        Construct a new object from `self` by replacing the attributes in `subs`.
```
Comment (georgebisbas) [inline devito/tools/abc.py:156]: OK

File: devito/arch/archinfo.py (hunk 2 above)
Comment (georgebisbas) [reply devito/arch/archinfo.py:23]: CPU64 = Cpu64('cpu64')

in archinfo

File: devito/operator/operator.py [hunk 6]
```diff
@@ -939,21 +955,23 @@ def _emit_apply_profiling(self, args):
...
+            # Same as above, but excluding the setup phase, e.g. the CPU-GPU
+            # data transfers in the case of a GPU run, mallocs, frees, etc.
+            metrics = []
+
+            v = summary.globals.get('fdlike-nosetup')
+            if v is not None:
+                metrics.append("%.2f s" % fround(v.time))
+                metrics.append("%.2f GPts/s" % fround(v.gpointss))
+
+                perf("Global performance <w/o setup>: [%s]" % ', '.join(metrics))
```
Comment (georgebisbas) [inline devito/operator/operator.py:967]: I tried to run locally the PR and I do not see any diff in the log, should I ?

File: devito/arch/archinfo.py (hunk 2 above)
Comment (mloubout) [reply devito/arch/archinfo.py:23]: Agrred, CPU64 + Cpu64 seems like there shound't need both

File: devito/operator/operator.py (hunk 6 above)
Comment (FabioLuporini) [reply devito/operator/operator.py:967]: it should yes according to these changes

File: devito/operator/operator.py (hunk 3 above)
Comment (FabioLuporini) [reply devito/operator/operator.py:958]: wdym ? this is just an additional line -- a line that does not include the setup phase

Comment (georgebisbas) [reply devito/operator/operator.py:958]: I will cross-check the logs again

File: devito/types/basic.py [hunk 7]
```diff
@@ -1247,6 +1248,19 @@ def _mem_mapped(self):
     def _mem_host(self):
         return self._space == 'host'
 
+    @cached_property
+    def _signature(self):
+        """
+        The signature of an AbstractFunction is the set of fields that
+        makes it "compatible" with another AbstractFunction. The fact that
```
Comment (EdCaunt) [inline devito/types/basic.py:1255]: Might be worth defining "compatible" here

File: devito/ir/clusters/algorithms.py [hunk 8]
```diff
@@ -433,7 +433,7 @@ def callback(self, clusters, prefix, seen=None):
 
             key = lambda i: i in prefix[:-1] or i in hs.loc_indices
             ispace = c.ispace.project(key)
-            # HaloTOuch are not parallel
+            # HaloTouch's are not parallel
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/ir/clusters/algorithms.py]: OK

File: devito/types/basic.py (hunk 7 above)
Comment (FabioLuporini) [reply devito/types/basic.py:1255]: that's encoded in the first part of the sentence

File: devito/operator/operator.py [hunk 9]
```diff
@@ -975,21 +993,12 @@ def lower_perfentry(v):
 
             if v.time <= 0.01:
                 # Trim down the output for very fast sections
-                name = "%s%s<>" % (k.name, rank)
+                name = "%s%s" % (k.name, rank)
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/operator/operator.py]: good catch, OK

File: devito/types/basic.py (hunk 7 above)
Comment (EdCaunt) [reply devito/types/basic.py:1255]: In what sense? The current docstring doesn't help me understand what it means for one `AbstractFunction` to be compatible with another in this context

**Review by EdCaunt — APPROVED** [review]

File: devito/types/basic.py (hunk 7 above)
Comment (FabioLuporini) [reply devito/types/basic.py:1255]: I will try to improve it in my next PR

**Pull request merged**